Cold regions harvester

joaandrade opened the pull request:
Fixes #

### Proposed fixes:
New harvester added to catalogue the Cold Regions Pilot outputs .
It is not supposed to add this harvester in CKAN. It is only required to run the python script (coldregions.py) in the command line, following the README instructions (or the [WIKI](https://github.com/NextGeoss/ckanext-nextgeossharvest/wiki/15.-Harvesting-Cold-Regions-pilot-outputs)).

`$ python ./src/ckanext-nextgeossharvest/ckanext/nextgeossharvest/harvesters/coldregions.py "http://thredds.nersc.no/thredds/catalog/nextgeoss/Svalbard_classification_2018/catalog.xml" <destination_ckan_URL> <destination_ckan_apikey> "nersc"`

If the Data Provider "nersc" does not exist, it will be created by the script. The datasets created by the script will also be added to the group "cold_regions". An ignore list was also added to avoid indexing of some metadata fields in SOLR (similar to the other harvesters).
By running this script, you will have 127 datasets for the collection "Sentinel-1 HH/HV based ice/water classification"

The following code line (necessary to run the extension with docker-compose) was added to the [plugin.py](https://github.com/NextGeoss/ckanext-nextgeossharvest/blob/cold-regions-harvester/ckanext/nextgeossharvest/plugin.py#L12) file:
`plugins.implements(plugins.IConfigurer, inherit=True)`

### Features:

- [X] includes new  features
- [ ] includes tests covering changes
- [X] includes updated documentation
- [ ] includes user-visible changes
- [ ] includes bugfix

Please [X] all the boxes above that apply

georgiana-b commented on ckanext/nextgeossharvest/harvesters/coldregions.py:
```diff
@@ -0,0 +1,216 @@
+	_close_session(session)
+
+except Exception:
+	print help_msg
```
Catching `Exception` here completely conceals any error that might occur from accessing the source and from processing the results, which may be completely unrelated to how the user was running the command. Doing this would give the user the impression that they're using the wrong command/parameters even when the problem may lie somewhere else in the code.

I think in this case the most sensible way would be to allow the errors to be thrown and the script to stop. The user can then get a traceback and fix them.

joaandrade replied:
Solved with the commit https://github.com/NextGeoss/ckanext-nextgeossharvest/pull/205/commits/7754ee33c6373822bfbaf01f75d16eb24b0f2ff0

georgiana-b commented on ckanext/nextgeossharvest/harvesters/coldregions.py:
```diff
@@ -0,0 +1,216 @@
+        ckan_url = sys.argv[2]
+        apikey = sys.argv[3]
+        owner_org = sys.argv[4]
+	harvest_url = sys.argv[1]
```
If you use `try` here to validate the command line arguments, please use the `except` right after this line to return the help message.

joaandrade replied:
In fact the error may appear on the two lines below. For example if the URL is wrong the line:

> r = request.get(harvest_url)

will have an error

georgiana-b replied:
Ok, but if that's the case the user will see it and resolve it accordingly. I think the responsibility of the script is just to ensure all the required parameters are being passed. The user has to make sure they are also correct.

joaandrade replied:
I understand you point. Solved with https://github.com/NextGeoss/ckanext-nextgeossharvest/pull/205/commits/69414c7c90448083a872d7fc46468188d209878f

georgiana-b commented on ckanext/nextgeossharvest/harvesters/coldregions.py:
```diff
@@ -0,0 +1,216 @@
+        }]
+
+
+# MAIN
```
Would be nice to go the standard way of [running scripts](https://gist.github.com/georgiana-b/e94b40f8a451096ecb3f655bed7f37e7#file-create_lai_product_datasets-py-L170).

joaandrade replied:
Solved with the commit https://github.com/NextGeoss/ckanext-nextgeossharvest/pull/205/commits/7754ee33c6373822bfbaf01f75d16eb24b0f2ff0

georgiana-b commented on ckanext/nextgeossharvest/harvesters/coldregions.py:
```diff
@@ -0,0 +1,216 @@
+
+	identifier = "NERSC_ARCTIC_SEAICEEDGE_" + time.replace("-","").replace(":","").replace(".000Z","")
+
+        extras_dict = {'collection_name': 'Sentinel-1 HH/HV based ice/water classification', 'collection_id': 'S1_ARCTIC_SEAICEEDGE_CLASSIFICATION', 'collection_description': 'Sea ice and water classification in the Arctic using Sentinel-1. These datasets have been generated with support from the European Commission in the Horizon 2020 NextGEOSS project (grant agreement No 730329). The algorithms ...
```
It seems very brittle to hardcode the values in here like this. Can't this information be extracted directly from the XML? I assume that's also where you retrieved it from.

joaandrade replied:
The information hardcoded such as collection name is not in the XML. It was discussed with the provider. The XML only lists the files within thredds.
In any case, this script is only for the cold regions outputs. We have collection names hardcoded in all the harvesters

georgiana-b replied:
Hey @joaandrade. I meant that about the `creator_name`, `source`, `contributors name`, etc. which could be subject to change?

joaandrade replied:
Hi @georgiana-b all of those fields are not in the XML. I opended a netcdf to check those values and then I confirmed with the provider that they are the same to all the products.
Only the spatial and times are different for each dataset. I was able to get each spatial from the WMS URL of each product and the time from the filename but nothing else. The XML is nothing more than a list of files.

georgiana-b replied:
I see. Thank you for the explanation!

joaandrade commented:
@georgiana-b swp files were removed at https://github.com/NextGeoss/ckanext-nextgeossharvest/pull/205/commits/582f39008f4b4e8a58053183f2c326ad903df7f5

joaandrade commented on ckanext/nextgeossharvest/harvesters/coldregions.py:
```diff
@@ -0,0 +1,216 @@
+    #groups = package_dict['groups']
+
+    headers = {
+        'Content-Type': 'application/x-www-form-urlencoded',
```
I'm building the system to test your suggestion. I will give you feedback soon

joaandrade replied:
Replaced at commit https://github.com/NextGeoss/ckanext-nextgeossharvest/pull/205/commits/9836017a2ac51f289953d49f6cdc898e77e39fca

joaandrade commented:
@georgiana-b all the comments wre solved with new commits or responded.
Right now, you won't find any datasets if you run the script because the data source (thredds) is down. I already contacted the data provider and I'll let you know as soon as it is up and running again.

joaandrade commented:
@georgiana-b the data source is now up and running so you can now test.
However, I would say to NOT merge since the provider told us about two extra collections that I will add between today and tomorrow to the script. The logic will be the same. I will just add the collection to harvest as an argument to be provided in the command. Is this ok for you?

georgiana-b commented:
Sure, whenever you're ready.

joaandrade commented:
In any case feel free to test for the current collection to check if the problems raised are solved.

georgiana-b commented:
Ok.
I also pushed the chages to `.gitignore` to [the other branch](https://github.com/NextGeoss/ckanext-nextgeossharvest/pull/206/commits/935d3f43e46285ba5de7ce14f0125a369cb1c0cd) just to make sure no more text editor files will be pushed by mistake.

joaandrade commented:
@georgiana-b the script was updated to include two more collections. Also the command to run the harvester does not include anymore the data source URL. Now use the collection ID. Find the instructions on README or WIKI (https://github.com/NextGeoss/ckanext-nextgeossharvest/wiki/15.-Harvesting-Cold-Regions-pilot-outputs)

The corrections we did this morning are still valid.
1st collection - 127 datasets
2nd collection - 200 datasets
3rd collection - 333 datasets

georgiana-b commented:
@joaandrade If the command line parameters are sent without quotes the script fails later on, in `_organization_list` which would make it difficult for the admin to figure out what is wrong.

Since we don't know who will end up running this in the future, I suggest parsing the argument with [`shplex.split`](https://docs.python.org/3/library/shlex.html#shlex.split) so that we don't have to worry about quotes.

joaandrade commented:
> @joaandrade If the command line parameters are sent without quotes the script fails later on, in `_organization_list` which would make it difficult for the admin to figure out what is wrong.
> 
> Since we don't know who will end up running this in the future, I suggest parsing the argument with [`shlex.split`](https://docs.python.org/3/library/shlex.html#shlex.split) so that we don't have to worry about quotes.

@georgiana-b the sys.argv is already an array with the arguments. How do you suggest to split it? Probably it is enough to ensure that all the arguments are strings using str()?!

georgiana-b commented:
@joaandrade `shlex.split` just makes sure to parse the command line parameters properly, regardless of how they are quoted.
Here is an example:

```
>>> import shlex
>>> shlex.split('unquoted parameters passed')
['unquoted', 'parameters', 'passed']
>>> shlex.split('"quoted" "parameters" "passed"')
['quoted', 'parameters', 'passed']
```

So even just using `shlex.split('  '.join(sys.argv[1:]))` would be enough to take care of all those cases.

Alternatively, we could go by the book with [`argparse`](https://docs.python.org/3/library/argparse.html#example) and define each parameter.
But we already know what kind of args we can get, I think the `shlex` solution is enough to cover all the cases.

joaandrade commented:
@georgiana-b agrs parsing implemented with the commit https://github.com/NextGeoss/ckanext-nextgeossharvest/commit/787cc84e81a586c8c722f035a0ceeab55b7d94f4

georgiana-b commented:
Perfect. Thanks a lot @joaandrade! 

I tested this on staging, including the new args parsing and it all works as stated so this is good to go.
